### ISSUE TITLE
Some styles can't be changed with imported editor themes

### ISSUE THREAD

**Issue by ghost** [godotengine/godot#8588]
Godot 3.0 30th april build linux x64 Kubuntu 17.04 

Changing resource for PanelContainer class does not change editor theming like it was doing in previous versions. 
STR: Make a theme, assign it as editor theme and try changing PanelContainer properties

![spectacle wl5593](https://cloud.githubusercontent.com/assets/10696955/25560207/4d2bb272-2d56-11e7-810f-ae3bbd632454.png)

ItemList class bg property is broken the same way but bg focus works fine

Button also gets inherited by all sorts of stuff and if you add any change to it's style items classes like CheckBox and switch in color picker get messed up. Changing their values to space out text and icon doesn't do anything.

**Comment by toger5** [comment]
There were some changes recently where some style boxed got moved to the editor them (instead default theme). Maybe if you check the editor theme class you will find a solution.
Just a guess though.

p.s. the colors are nice ;) really nice idea to blend it in with your system window color!

**Comment by ArcFutahito** [comment]
Hell yeah for breeze dark experience! :D
![spectacle wz4338](https://cloud.githubusercontent.com/assets/10696955/25561819/35fb7f26-2d7c-11e7-8bb1-5d9bb43e994e.png)
Got most of the stuff working but that button inheritance and panel container are just ugh. Editor class is all about script editor, don't see anything that would point to panels.

**Comment by toger5** [comment]
Holy ***
good job. I actually really like the text inputs you designed!
let me search the file real quick I was referring...

**Comment by toger5** [comment]
okay here we go: https://github.com/godotengine/godot/blob/2153feb6fd8cc328191c92b43818624a6fa6c27e/editor/editor_themes.cpp

this file also introduces some style boxes. which are independent of the default theme. And it seems like it is setting custom colors: It uses styleBoxFlat so the textures of course have no impact...

**Comment by ArcFutahito** [comment]
Yeah I suspected that some of the stuff got hardcoded into godot during theme transition and I can't wire up to it from tres. I guess it's just plain impossible to finish theme for now. 

If you wanna play around with it, here's the file. Also, full disclosure, breeze is work of kde people, I'd never come up with something so sexy :D
[breeze_dark.tres.tar.gz](https://github.com/godotengine/godot/files/966595/breeze_dark.tres.tar.gz)

**Comment by ArcFutahito** [comment]
Also I have no idea how to give tres relative paths to icons so it would work, for example, with icons that are located inside theme folder. I'd pump it full of sweet breeze stuff if I could get those to work in a nice drag n droppable portable package.

**Comment by toger5** [comment]
oh you didn't even touch the editor_theme.cpp ?? it's just a imported theme. Even nicer.
Maybe you should rephrase the issue description to: `Some styles can't be changed with imported editor themes.` because that is actually an issue ;)
@volzhs I think this is because of the changes to the theme in the file I linked before. do you have an idea how to fix it so that imported editor themes can overwrite it?

Just checked the file again. maybe it is not caused by the changes you made. since imported themes are custom themes as far as I know so they use the other function...
@volzhs maybe you have another an idea

**Comment by toger5** [comment]
@ArcFutahito icons are not handled by the theme. they are part of the editor itself. so you can't change without recompile...
(also something to take into consideration... customisable icon sets. But I really like the current one)

**Comment by ArcFutahito** [comment]
They actually ARE handled by the theme. It's just that paths are not relative and you need them kicking around in res:// to get loaded

**Comment by toger5** [comment]
can u post a screenshot of the messed up color picker (when assigning a style box for buttons)

**Comment by ArcFutahito** [comment]
![spectacle ns5913](https://cloud.githubusercontent.com/assets/10696955/25562426/af430630-2d8e-11e7-946b-aa771f1e6f38.png)
Default color picker. Messed up but I've seen some tickets around. 
![spectacle lh5972](https://cloud.githubusercontent.com/assets/10696955/25562474/4b89cf74-2d8f-11e7-985d-41db87b932ff.png)
After restyling buttons this happens to RAW mode checkbox button and other controls that inherit Button. I've been digging around affected classes trying to straighten it out but nothing seems to work.

**Comment by toger5** [comment]
@ArcFutahito What icons? Are you talking about images for style boxes, or icons for buttons: search icon, expand icon ... or node icons.

Style box textures are supported. For editor icons I think only some are accessible but I'm not sure. (for example the expand in trees could be part of the theme) but I thinks the search icon for example is not part of the theme file. and the node icons aren't either. If I'm wrong I'm happy to know!

Yea and the path for the style box textures is an issue... someone once posted a theme which had an install file than adapted all the path's in the theme file... 
Another topic that could need an issue... maybe I can link you the github when I find it.

**Comment by ArcFutahito** [comment]
All icons ARE handled by theme. They are a bitch to get to because reduz decided to make it easter egg or something. Had to colorize screenshot because making empty template fucks everything up but here's a fraction of stuff kicking around dusty corners.
![sc-10_29_37-3004](https://cloud.githubusercontent.com/assets/10696955/25562513/20ef626e-2d90-11e7-9a7a-8853eab00084.png)

**Comment by toger5** [comment]
there we go: https://github.com/Geequlim/godot-themes
qeequlim had a install script check it out. that could make It work... although there should be a better solution without install script. but with relative file paths...

**Comment by toger5** [comment]
@ArcFutahito holy.. okay thats really nice to know. Need to check it out immediately Thank you!

**Comment by ArcFutahito** [comment]
![images duckduckgo com](https://cloud.githubusercontent.com/assets/10696955/25562551/2e8a27b4-2d91-11e7-8efa-48e6ff5b1a97.jpeg)
Empty editor template adds a whole metric ton of classes not accessible from dropdown because reasons. Last time I checked, everything worked just fine.

**Comment by toger5** [comment]
@ArcFutahito you are totally right sorry! Never was brave enough to press `Create Empty Editor Template` ;)
So it's just about making the paths relative and than we will get flooded with themes

**Comment by toger5** [comment]
Interesting you are right they are only accessible through the editor theme button. which makes sense to be honest. Most of those values are irrelevant for any other theme! I like that. although it's hard to find.
Did you check out the install script?

**Comment by ArcFutahito** [comment]
Yeah, making those avaliable from "Add Class" menu would be nice.

**Comment by ArcFutahito** [comment]
Checking it out right now.

**Comment by ArcFutahito** [comment]
Theme resources are kicking in res:// in there too. For example: res://addons/arc_dark/resources/button/button-insensitive.tres

**Comment by toger5** [comment]
in the [install Script](https://github.com/Geequlim/godot-themes/blob/master/install.py)
line 33
`content = content.replace("res://addons", themdir)`
he is changing the actual theme.tres file to adapt all the paths...
this is how it works for his theme but I would consider that quite hacky...

**Comment by toger5** [comment]
`../resources/button/button-insensitive.tres`
this should be supported for imported themes. relative to the path passed to the custom theme property.

**Comment by toger5** [comment]
oh and the theme generation needs to have an option for relative paths...

**Comment by ArcFutahito** [comment]
Well, icons are not a priority yet anyway, at least not as long as there are random hardcoded panels lying around

**Comment by toger5** [comment]
I think I found the reason: since volzhs's update to the theme the main editor uses the EditorPanel style box from EditorStyles
`bottom_panel->add_style_override("panel", gui_base->get_stylebox("EditorPanel", "EditorStyles"));`
the reason is, that it needs a adapted size...

Ill check if that is accessible through the theme creator.

**Comment by toger5** [comment]
just found this, maybe thats interesting:
```
Ref<Theme> theme = create_editor_theme();
theme_base->set_theme(theme);
gui_base->set_theme(create_custom_theme());
```
first the default theme gets added to theme_base, than the custom theme (usually empty) is to the gui_base. All Childs use the theme from the closest parent, so when you give it a style for a specific element it uses it, for any undefined elements it just falls back to the default editor theme. because thats is assigned to theme_base node.
In your theme panel is set, so before volzhs changes it used your theme. not the editor panel doesn't search anymore for panel, instead it searches for EditorPanel which is not set in your theme so if falls back to the blue design.

Most of it is probably not new for you but I still wanted to explain it in detail.
A fix would be to add the EditorPanel to the list in the theme editor.

**Comment by toger5** [comment]
just checked some more code and basically it should be there. I'm currently recompiling and check if I can find the Style.

**Comment by toger5** [comment]
<img width="1680" alt="screen shot 2017-04-30 at 01 55 42" src="https://cloud.githubusercontent.com/assets/16718859/25563060/6e2cbe98-2d48-11e7-98b3-f4514acd5aaf.png">
Here we go!
So everything works as intended, it's just a little bit confusing, that it is not using the default panel. but it also has advantages in terms of Customisation...
So this one is solved but I thin we should open another issue: `make custom theme paths relative for easier installation.`
other option would be to create a custom binary type which includes the assets with the theme...

**Comment by ArcFutahito** [comment]
Thanks a lot! I got way too focused on how previous versions of editor were working and that was my undoing.
Got those panels working, they also didn't update without restarting the editor so it made everything even more complicated.

**Comment by ArcFutahito** [comment]
So godot just decided to wipe the crap out of the entire tres. Fun times, ha ha!

**Comment by toger5** [comment]
This looks really good. some ppl might even prefer it as the default there is a color argument going on... So maybe as soon as the installation process becomes easier, this could be a unofficial official alternative or even shipped with the engine (since you don't have any textures your theme is tiny).

maybe we also can investigate into the issue with the dock borders you were bringing up.
Can you rephrase it so I know what exactly you expected and what is not working?

**Comment by toger5** [comment]
@ArcFutahito: "Kinda sad couldn't get sweet borders around tabs to be displayed like in default widgets =\"
Is it a technical limitation or can you not get it so that there is a gap in the border around the tabContainer so that it looks wired?
If that is the case I have an idea. you can use expands to make the tabs overlap down over the tab container. and than the style box of the tab's has no border at the bottom but it has borders around the the top part. the tab container has borders all around but they are hidden underneath the bottom expand from the active tab.
It's just a guess and maybe you had a different issue...

**Comment by ArcFutahito** [comment]
>Is it a technical limitation or can you not get it so that there is a gap in the border around the tabContainer so that it looks wired?

Technical limitation, tab borders are your plain old rectangles as well as tab containers. I can't control border drawing separately because there was a genius idea to make two border options: left+top and right+bottom. 
Since you can't just disable one of the borders tab title looks like a rectangle slapped on top of another rectangle. 

It's not exactly a HUGE thing but having more control over stylebox borders would be definitely a nice thing to have. 

> this could be a unofficial alternative

That's the idea.

**Comment by ArcFutahito** [comment]
Basically, what I think the reason for this is how polygons get drawn on the low level. If my hunch is correct it probably looks like the picture on top and that's why you can control two sides in a pack deal. Making styleboxes drawn with four polys would allow to control each border separately.
![spectacle ns7862](https://cloud.githubusercontent.com/assets/10696955/25568606/155e0200-2e0f-11e7-9c9e-9fc383cf3b9a.png)

But I'm speaking out of my ass so honestly it's just a speculation

**Comment by MakeGodotGreatAgain** [comment]
god, the buttons on that theme is absolutely horrendous. hope that shit gets fixed before launch

**Comment by toger5** [comment]
@ArcFutahito Okay I see what you mean. So you could achieve it with texture style boxes. but there for you need the relative paths to make it easier to install.
I also really like the idea of giving an option for rounded corners for flat styleboxes that would make it flat boxes sooo useful:
perfect would be:
```
roundedTL: Float (radius)
roundedTR: Float (radius)
roundedBL: Float (radius)
roundedBR: Float (radius)
ColorTL: Color
ColorTR: Color
ColorBL: Color
ColorBR: Color
WidthTL: Float 
WidthTR: Float 
WidthBL: Float 
WidthBR: Float 
fillColor: Color
```
@MakeGodotGreatAgain I dont know what you mean? this thread is not about any theme launch.
It is also not about the default editor theme??
Or are you talking about the default editor theme? than this is the wrong thread...

**Comment by ArcFutahito** [comment]
>So you could achieve it with texture style boxes. but there for you need the relative paths to make it easier to install.

Personally I would try to omit using textures for something that doesn't require a lot of resources to get rendered especially if functionality can be used for a whole lot of different purposes (at least I THINK people would like to be able to control borders separately and have rounded corners for their game UI needs). 
It's doable with textures though, yes.

**Comment by ArcFutahito** [comment]
![spectacle jd3421](https://cloud.githubusercontent.com/assets/10696955/25586015/0d49d8ae-2ea6-11e7-9c80-f366fa40c095.png)
Completion panel edit is for panel on background of completion. There seems to be no way to change actual completion window.

**Comment by volzhs** [comment]
![godot3](https://cloud.githubusercontent.com/assets/8281454/25586378/12841044-2eda-11e7-9f87-fcff7babd225.gif)

here's my WIP.
what I'm trying to do is that theme color can be changed in editor setting.

**Comment by ArcFutahito** [comment]
Were you able to do anything with CheckBox class? I'm trying to change text margin but none of the parameters are working. 
![spectacle qv3421](https://cloud.githubusercontent.com/assets/10696955/25586934/5fa5c492-2eaa-11e7-8cbd-fcbd6730b80c.png)

**Comment by volzhs** [comment]
I'm doing one by one for better look.
I will take a look later.

**Comment by ArcFutahito** [comment]
Please tell me if you find anything, those checkboxes are driving me nuts.

**Comment by toger5** [comment]
@ArcFutahito I think almost all the things we addressed in this discussion are solved.
You could consider closing I think.
even the borders you wanted to get the same than they are in Breeze could work. since you now can do individual border width + rounded corners.
you could also add a expand margin to the bottom of the tabs so that there is a connection without line to the container.

**Comment by volzhs** [comment]
I guess Custom Theme is useless now at editor settings > interface > theme.
![image](https://user-images.githubusercontent.com/8281454/29933590-2474925e-8eb3-11e7-82dc-9b59649563fb.png)
could we remove it?

**Comment by toger5** [comment]
@volzhs I wouldn't remove since it still could be cool when wanting to create icon themes.. or when someone wants something special with less modern rounder...

**Comment by kubecz3k** [comment]
First of all thank you for your report and sorry for the delay.

We released Godot 3.0 in January 2018 after 18 months of work, fixing many old issues either directly, or by obsoleting/replacing the features they were referring to.

We still have hundreds of issues whose relevance/reproducibility needs to be checked against the [current stable version](https://godotengine.org/download), and that's where you can help us.
**Could you check if the issue that you described initially is still relevant/reproducible in Godot 3.0 or any newer version, and comment about its current status here?**

For bug reports, please also make sure that the issue contains detailed steps to reproduce the bug and, if possible, a zipped project that can be used to reproduce it right away. This greatly speeds up debugging and bugfixing tasks for our contributors.

Our Bugsquad will review this issue more in-depth in 15 days, and potentially close it if its relevance could not be confirmed.

Thanks in advance.

*Note: This message is being copy-pasted to many "stale" issues (90+ days without activity). It might happen that it is not meaningful for this specific issue or appears oblivious of the issue's context, if so please comment to notify the Bugsquad about it.*

**Comment by akien-mga** [comment]
Fixed by #8631.
